torch/distributions/transforms.py

```diff
@@ -0,0 +1,342 @@
+    det jacobians. They are primarily used in
+    :class:`torch.distributions.TransformedDistribution`.
+
+    Caching is useful for tranforms whose inverses are either expensive or
+    numerically unstable. Note that care must be taken with memoized values
+    since the autograd graph may be reversed. For example while the following
+    works with or without caching::
+
+        y = t(x)
+        t.log_abs_det_jacobian(x, y).backward()  # x will receive gradients.
+
+    However the following will error when caching due to dependency reversal::
+
+        y = t(x)
+        z = t.inv(y)
+        grad(z.sum(), [y])  # error because z is x
+
+    Derived classes should implement one or both of :meth:`_call` or
+    :meth:`_inverse`. Derived classes that set `bijective=True` should also
+    implement :meth:`log_abs_det_jacobian`.
+    """
+    bijective = False
+
+    def __init__(self, cache_size=0):
+        if cache_size != 0:
+            if cache_size == 1:
+                self._cached_x_y = None, None
+                self.__call__ = self._cached_call
+                self.inverse = self._cached_inverse
+            else:
+                raise NotImplementedError('cache_size must be 0 or 1')
+
+    @lazy_property
+    def inv(self):
+        """
+        Returns the inverse :class:`Transform` of this transform.
+        """
+        return InverseTransform(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    def __ne__(self, other):
+        # Necessary for Python2
+        return not self.__eq__(other)
+
+    def __call__(self, x):
+        """
+        Computes the transform `x => y`.
+        """
+        return self._call(x)
+
+    def inverse(self, y):
+        """
+        Inverts the transform `y => x`.
+        """
+        return self._inverse(y)
+
+    def _cached_call(self, x):
+        """
+        Computes the memoized transform `x => y`.
+        """
+        x_old, y_old = self._cached_x_y
+        if x is x_old:
+            return y_old
+        y = self._call(x)
+        self._cached_x_y = x, y
+        return y
+
+    def _cached_inverse(self, y):
+        """
+        Inverts the memoized transform `y => x`.
+        """
+        x_old, y_old = self._cached_x_y
+        if y is y_old:
+            return x_old
+        x = self._inverse(y)
+        self._cached_x_y = x, y
+        return x
+
+    def _call(self, x):
+        """
+        Abstract method to compute forward transformation.
+        """
+        raise NotImplementedError
+
+    def _inverse(self, y):
+        """
+        Abstract method to compute inverse transformation.
+        """
+        raise NotImplementedError
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Computes the log det jacobian `log |dy/dx|` given input and output.
+        """
+        raise NotImplementedError
+
+
+class InverseTransform(Transform):
+    """
+    Inverts a single :class:`Transform`.
+    """
+    __slots__ = ['inv']

```

Comment:
Are slots necessary here? The downside is that users won't be able to assign custom attributes to transforms (not sure how important that is, but it might turn out to be useful at some point). I don't think allocating a `__dict__` would make a big difference. Also, these transforms won't support weak refs

Reply:
Thanks, I didn't know about these disadvantages. Removed.